### ISSUE TITLE
Inferred type for a variable annotated as a union of parameterized lists is incorrect when assigned an empty list

### ISSUE THREAD

**Issue by Sachaa-Thanasius** [microsoft/pyright#8111]
**Describe the bug**
If a variable is annotated as a parameterized list, e.g. `list[int] | list[str]`, but assigned an empty list, pyright will infer the type of the variable to be the only part of the union and ignore the rest of the annotation.

**Code or Screenshots**

```python
from typing_extensions import reveal_type

# Expected types for all of these should match the original annotations, I think.

inp: list[str] | list[int] = []
reveal_type(inp)  # Type of "inp" is "list[int]"

inp2: list[str] | list[tuple[str, object]] = []
reveal_type(inp2)  # Type of "inp2" is "list[str]"

inp3: list[str] | list[tuple[str, object]] | list[int] | list[float] = []
reveal_type(inp3)  # Type of "inp3" is "list[float]"
```

**VS Code extension or command-line**
Running pyright via Pylance (version 2024.6.1) in VSCode.

**Comment by Sachaa-Thanasius** [comment]
**Follow-up**

Here's a case where pyright prioritizes its own inference over the intentionally wide annotation when an assignment is involved. Might be that this is a more general issue about inference priority.

```py
from collections.abc import Sequence

from typing_extensions import reveal_type

def example(member_names: Sequence[str] | Sequence[tuple[str, object]]):
    reveal_type(member_names)  # As expected: Type of "member_names" is "Sequence[str] | Sequence[tuple[str, object]]"

    names_seq: list[str] | list[tuple[str, object]] | list[str | tuple[str, object]] = list(member_names)

    reveal_type(names_seq)  # Unexpected: Type of "names_seq" is "list[str | tuple[str, object]]"
```

**Comment by erictraut** [comment]
Pyright is working as intended here, so this isn't a bug.

An object's specialized type is established at the time it is constructed. Let's look at your first example, which involves the statement `inp: list[str] | list[int] = []`. This statement has two parts. In fact, it might be easier to tell what's happening here if we write it as two separate statements:

```pyright
inp: list[str] | list[int]
inp = []
```

First, the variable `inp` is declared with a type of `list[str] | list[int]`. That means pyright will enforce that any subsequent assignments to this variable must be compatible with `list[str] | list[int]`. Second, you are assigning the value `[]` to `inp`. The expression `[]` constructs a new list object. The specialized type of an object must be established at the time it is constructed. Pyright can tell from the declared type of `inp` that you intend for this constructed list to either be `list[str]` _or_ `list[int]`. Its type cannot be both `list[str]` _and_ `list[int]`. It must choose one of the two types for this constructed object. Regardless of which type pyright chooses for the expression `[]`, this value is then assigned back to `inp`, and the variable's type is narrowed on assignment.

```python
inp: list[str] | list[int]
inp = []
```

For more information, refer to the documentation on [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types) and [list expressions](https://microsoft.github.io/pyright/#/type-inference?id=list-expressions).

**Comment by erictraut** [comment]
> I agree that pyright shouldn't think inp is both a list of strings and a list of integers

The question is: "what type is `[]` at the time it's constructed?". This is the critical decision. There are two obvious choices given the bidirectional inference context: `list[int]` and `list[str]`. It can't be both, so pyright needs to decide between them. That decision should be deterministic (i.e. not vary from one run to the next) and should not depend on union ordering. Pyright makes the decision by first "sorting" all of the subtypes in the union using a deterministic (but undocumented) sorting algorithm. It then filters any subtypes that are not applicable. (For example, if the inference context is `list[str] | tuple[int]`, it can eliminate `tuple[int]`.) If two or more candidates remain, it picks the first one.

> I still don't get why it has to choose between them in your example

Because the operations that can be performed on that variable subsequent to the assignment depend on its type. If the type of `[]` is inferred to be `list[int]`, then it's valid to call `imp.append(1)` and invalid to call `imp.append("a")`. If the type is inferred to be `list[str]`, then the converse is true. If the type is inferred to be `list[int] | list[str]`, then neither of those `append` calls would be allowed because they would both violate typing rules.
